### ISSUE TITLE
write is locale-dependent

### ISSUE THREAD

**Issue by Oxyd** [ashinn/chibi-scheme#829]
Reproducer:
```c
#include <chibi/eval.h>
#include <locale.h>

int
main() {
  setlocale(LC_ALL, "cs_CZ.UTF-8");

  sexp ctx;
  sexp_scheme_init();
  ctx = sexp_make_eval_context(NULL, NULL, NULL, 0, 0);
  sexp_load_standard_env(ctx, NULL, SEXP_SEVEN);
  sexp_load_standard_ports(ctx, NULL, stdin, stdout, stderr, 1);

  sexp_eval_string(ctx, "(write 2.3)", -1, NULL);

  sexp_destroy_context(ctx);
  return 0;
}
```
Observed output:
```
% ./chibi-repro 
2,3.0
```
This is because Chibi uses `printf` to format floating-point values, but `printf` produces locale-dependent output and the cs_CS locale, for example, uses a comma as the decimal separator instead of a dot as Chibi expects.

An obvious workaround would be for Chibi to temporarily change the locale before using `printf`, however, this introduces another issue: `setlocale` isn't thread-safe and a program can't make calls to `setlocale` in parallel with anything that might be writing or using the current locale, such as other I/O functions. Since Chibi is meant to be used as a library, the code should be prepared to be embedded in multi-threaded programs, so it shouldn't introduce data races like that.

`uselocale` might be an option instead, but it's not standard C, so it creates portability issues.

**Comment by ashinn** [comment]
Cute.  I'm surprised no one has found this before.

The trailing `.0` is from a post-format check to enforce inexact formatting.  We can at the same time check for a comma and convert to decimal, which in practice should be the only problem.

**Comment by Oxyd** [comment]
Not sure that's the only problem.
```
% LC_ALL=ps_AF.UTF-8 printf "%g\n" 3.2
3٫2
```
That's U+066B 'ARABIC DECIMAL SEPARATOR'.

Also some languages use an entirely different set of numerals. I believe that on Linux glibc doesn't support locales with different numerals, but other OSs might. OSX in particular would be the prime suspect here, but I don't have a Mac to test with.

**Comment by ashinn** [comment]
I can't actually reproduce this with env vars:

`LANG=cs_CZ.UTF-8 LC_ALL=cs_CZ.UTF-8 LC_NUMERIC=cs_CZ.UTF-8 chibi-scheme -p3.2`

just outputs 3.2 on my Linux machine (which is probably why noone has reported this - it doesn't apply to the chibi-scheme main).  However

`LC_NUMERIC=cs_CZ.UTF-8 /usr/bin/printf "%g\n" 3.2`

outputs `3,2` (note the bash builtin printf expects the read format for numbers to match the locale, unlike the /usr/bin binary).

I can reproduce your C example, though, so glibc is treating the env vars differently from an explicit setlocale request.

On Mac I can't get it to print any alternate numerals.

**Comment by lassik** [comment]
Good catch. The same problem could exist in other Scheme implementations.

Might be easier to import a locale-independent float printer from somewhere than work around all the gotchas?

**Comment by lassik** [comment]
Or maybe not, it's [quite a bit of subtle code](https://github.com/jwiegley/gdtoa).

**Comment by ashinn** [comment]
I implemented accurate floating point formatting from scratch in `fmt`, it's quite complicated and not something I want to pull into the Chibi core.  Let me try a simple workaround.
